drivers/serial/uart_nrfx_uart.c

```diff
@@ -291,6 +318,344 @@ static int uart_nrfx_config_get(struct device *dev, struct uart_config *cfg)
 	return 0;
 }
 
+
+#ifdef CONFIG_UART_ASYNC_API
+
+static void user_callback(struct uart_event *event)
+{
+	if (uart0_cb.callback) {
+		uart0_cb.callback(event, uart0_cb.user_data);
+	}
+}
+
+static int uart_nrfx_callback_set(struct device *dev, uart_callback_t callback,
+				  void *user_data)
+{
+	uart0_cb.callback = callback;
+	uart0_cb.user_data = user_data;
+
+	return 0;
+}
+
+static int uart_nrfx_tx(struct device *dev, u8_t *buf, size_t len,
+			u32_t timeout)
+{
+	if (uart0_cb.tx_buffer_length != 0) {
+		return -EBUSY;
+	}
+
+	uart0_cb.tx_buffer = buf;
+	uart0_cb.tx_buffer_length = len;
+#if	defined(DT_NORDIC_NRF_UART_UART_0_RTS_PIN) && \
+	defined(DT_NORDIC_NRF_UART_UART_0_CTS_PIN)
+	uart0_cb.tx_timeout = timeout;
+#endif
+	nrf_uart_event_clear(uart0_addr, NRF_UART_EVENT_TXDRDY);
+	nrf_uart_task_trigger(uart0_addr, NRF_UART_TASK_STARTTX);
+	nrf_uart_event_clear(uart0_addr, NRF_UART_EVENT_TXDRDY);
+	nrf_uart_int_enable(uart0_addr, NRF_UART_INT_MASK_TXDRDY);
+
+	u8_t txd = uart0_cb.tx_buffer[uart0_cb.tx_counter];
+
+	nrf_uart_txd_set(uart0_addr, txd);
+
+	return 0;
+}
+
+static int uart_nrfx_tx_abort(struct device *dev)
+{
+	if (uart0_cb.tx_buffer_length == 0) {
+		return -EFAULT;
```

Comment:
Description of EFAULT:

> #define EFAULT          14      /* Bad address */

More specifically, EFAULT is returned when some pointer arg to a syscall is invalid and dereferencing it leads to CPU. That's pretty bad error code to use here. Unfortunately, there doesn't seem to be good POSIX error code for this case ("Invalid state for requested operation"). But EINVAL is still better than EFAULT.